File: plugin/src/main/java/com/stratio/cassandra/lucene/index/DocumentIterator.java
/*
 * Copyright (C) 2014 Stratio (http://stratio.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.stratio.cassandra.lucene.index;

import com.stratio.cassandra.lucene.IndexException;
import com.stratio.cassandra.lucene.util.TimeCounter;
import com.stratio.cassandra.lucene.util.Tracer;
import org.apache.cassandra.utils.CloseableIterator;
import org.apache.cassandra.utils.Pair;
import org.apache.lucene.document.Document;
import org.apache.lucene.search.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.*;

/**
 * {@link Iterator} for retrieving Lucene {@link Document}s satisfying a {@link Query} from an {@link IndexSearcher}.
 *
 * @author Andres de la Pena {@literal <adelapena@stratio.com>}
 */
public class DocumentIterator implements CloseableIterator<Pair<Document, ScoreDoc>> {

    private static final Logger logger = LoggerFactory.getLogger(DocumentIterator.class);

    /** The max number of rows to be read per iteration. */
    static final int MAX_PAGE_SIZE = 10000;

    private final SearcherManager manager;
    private final Query query;
    final int page;
    private final Deque<Pair<Document, ScoreDoc>> documents = new LinkedList<>();
    private final Sort sort, indexSort;
    private final Set<String> fields;
    private ScoreDoc after = null;
    private boolean finished = false;
    private IndexSearcher searcher;
    private int numReadDocuments = 0;
    private final Query startQuery;

    /**
     * Builds a new iterator over the {@link Document}s satisfying the specified {@link Query}.
     *
     * @param manager the Lucene index searcher manager
     * @param query the query to be satisfied by the documents
     * @param sort the sort in which the documents are going to be retrieved
     * @param page the iteration page size
     * @param fields the names of the document fields to be loaded
     */
    DocumentIterator(SearcherManager manager,
                     Sort indexSort,
                     Query after,
                     Query query,
                     Sort sort,
                     int page,
                     Set<String> fields) {
        this.manager = manager;
        this.query = query;
        this.indexSort = indexSort;
        this.fields = fields;
        this.startQuery = after;
        this.page = Math.min(page, MAX_PAGE_SIZE) + 1;
        TimeCounter time = TimeCounter.create().start();
        try {
            searcher = manager.acquire();
            this.sort = sort.rewrite(searcher);
            if (after != null) {
                Tracer.trace("Searching for last Lucene index position");
                BooleanQuery.Builder builder = new BooleanQuery.Builder();
                builder.add(after, BooleanClause.Occur.FILTER);
                builder.add(query, BooleanClause.Occur.MUST);
                ScoreDoc[] scoreDocs = searcher.search(builder.build(), 1, this.sort).scoreDocs;
                if (scoreDocs.length > 0) {
                    this.after = scoreDocs[0];
                }
            }
        } catch (IOException e) {
            throw new IndexException(e, "Error while acquiring index searcher");
        } finally {
            logger.debug("Index query initialized in {}", time.stop());
        }
    }

    private synchronized void fetch() {

        try {

            TimeCounter time = TimeCounter.create().start();

            TopDocs topDocs;
            if (startQuery == null && EarlyTerminatingSortingCollector.canEarlyTerminate(sort, indexSort)) {
                FieldDoc fieldDoc = after == null ? null : (FieldDoc) after;
                TopFieldCollector collector = TopFieldCollector.create(sort, page, fieldDoc, true, false, false);
                int hits = numReadDocuments + page;
                searcher.search(query, new EarlyTerminatingSortingCollector(collector, sort, hits, indexSort));
                topDocs = collector.topDocs();
            } else {
                topDocs = searcher.searchAfter(after, query, page, sort);
            }

            ScoreDoc[] scoreDocs = topDocs.scoreDocs;
            numReadDocuments += scoreDocs.length;
            finished = scoreDocs.length < page;
            for (ScoreDoc scoreDoc : scoreDocs) {
                after = scoreDoc;
                Document document = searcher.doc(scoreDoc.doc, fields);
                documents.add(Pair.create(document, scoreDoc));
            }

            Tracer.trace("Lucene index fetches {} documents", scoreDocs.length);
            logger.debug("Index query page fetched with {} documents in {}", scoreDocs.length, time.stop());

        } catch (Exception e) {
            close();
            throw new IndexException(logger, e, "Error searching in with {} and {}", query, sort);
        }

        if (finished) {
            close();
        }
    }

    /**
     * Returns {@code true} if the iteration has more {@link Document}s. (In other words, returns {@code true} if {@link
     * #next} would return an {@link Document} rather than throwing an exception.)
     *
     * @return {@code true} if the iteration has more {@link Document}s
     */
    @Override
    public boolean hasNext() {
        if (needsFetch()) {
            fetch();
        }
        return !documents.isEmpty();
    }

    /**
     * Returns if more {@link Document}s should be fetched from the Lucene index.
     *
     * @return {@code true} if more documents should be fetched, {@code false} otherwise
     */
    public boolean needsFetch() {
        return !finished && documents.isEmpty();
    }

    /**
     * Returns the next {@link Document} in the iteration.
     *
     * @return the next document
     * @throws NoSuchElementException if the iteration has no more {@link Document}s
     */
    @Override
    public Pair<Document, ScoreDoc> next() {
        if (hasNext()) {
            return documents.poll();
        } else {
            throw new NoSuchElementException();
        }
    }

    /** {@inheritDoc} */
    @Override
    public synchronized void close() {
        if (searcher != null) {
            try {
                manager.release(searcher);
            } catch (IOException e) {
                throw new IndexException("Error while releasing index searcher");
            } finally {
                searcher = null;
            }
        }
    }
}

File: plugin/src/main/scala/com/stratio/cassandra/lucene/index/DocumentIterator.scala
/*
 * Copyright (C) 2014 Stratio (http://stratio.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.stratio.cassandra.lucene.index

import com.stratio.cassandra.lucene.IndexException
import com.stratio.cassandra.lucene.index.DocumentIterator._
import com.stratio.cassandra.lucene.util.{Logging, TimeCounter, Tracing}
import org.apache.cassandra.utils.CloseableIterator
import org.apache.lucene.document.Document
import org.apache.lucene.index.Term
import org.apache.lucene.search.BooleanClause.Occur._
import org.apache.lucene.search.EarlyTerminatingSortingCollector._
import org.apache.lucene.search._

/** [[CloseableIterator]] for retrieving Lucene documents satisfying a query.
  *
  * @param cursors   the searcher managers and pointers of the involved indexes
  * @param indexSort the sort of the index
  * @param querySort the sort in which the documents are going to be retrieved
  * @param query     the query to be satisfied by the documents
  * @param limit     the iteration page size
  * @param fields    the names of the document fields to be loaded
  * @author Andres de la Pena `adelapena@stratio.com`
  */
class DocumentIterator(
    cursors: List[(SearcherManager, Option[Term])],
    indexSort: Sort,
    querySort: Sort,
    query: Query,
    limit: Int,
    fields: java.util.Set[String])
  extends Iterator[(Document, ScoreDoc)] with AutoCloseable with Logging with Tracing {

  private[this] val pageSize = Math.min(limit, MAX_PAGE_SIZE) + 1
  private[this] val documents = new java.util.LinkedList[(Document, ScoreDoc)]
  private[this] val indices = cursors.indices
  private[this] val managers = cursors.map(_._1)
  private[this] val searchers = managers.map(_.acquire())
  private[this] val afterTerms = cursors.map(_._2)
  private[this] val offsets = cursors.map(_ => 0).toArray
  private[this] var finished = false
  private[this] var closed = false

  private[this] def releaseSearchers(): Unit =
    indices.foreach(i => managers(i).release(searchers(i)))

  /** The sort of the query rewritten by the searcher. */
  private[this] val sort = try {
    querySort.rewrite(searchers.head)
  } catch {
    case e: Exception =>
      releaseSearchers()
      throw new IndexException(e, s"Error rewriting sort $indexSort")
  }

  /** The start after positions. */
  private[this] val afters = try {
    indices.map(i => afterTerms(i).map(term => {
      val time = TimeCounter.start
      val builder = new BooleanQuery.Builder
      builder.add(new TermQuery(term), FILTER)
      builder.add(query, MUST)
      val scores = searchers(i).search(builder.build, 1, sort).scoreDocs
      if (scores.nonEmpty) {
        tracer.trace("Lucene index seeks last index position")
        logger.debug(s"Start position found in $time")
        scores.head
      } else throw new IndexException("Last page position not found")
    })).toArray
  } catch {
    case e: Exception =>
      releaseSearchers()
      throw new IndexException(e, "Error while searching for the last page position")
  }

  private[this] def fetch() = {
    try {
      val fetchTime = TimeCounter.start

      val fieldDocs = indices.map(i => {
        val afterTerm = afterTerms(i)
        if (afterTerm.isEmpty && canEarlyTerminate(sort, indexSort)) {
          val fieldDoc = afters(i).map(_.asInstanceOf[FieldDoc]).orNull
          val collect = TopFieldCollector.create(sort, pageSize, fieldDoc, true, false, false)
          val hits = offsets(i) + pageSize
          val earlyCollect = new EarlyTerminatingSortingCollector(collect, sort, hits, indexSort)
          searchers(i).search(query, earlyCollect)
          val topDocs = collect.topDocs
          offsets(i) += topDocs.scoreDocs.length
          topDocs
        } else searchers(i).searchAfter(afters(i).orNull, query, pageSize, sort, false, false)
      }).toArray

      // Merge partitions results
      val scoreDocs = TopDocs.merge(sort, pageSize, fieldDocs).scoreDocs

      val numFetched = scoreDocs.length
      finished = numFetched < pageSize

      for (scoreDoc <- scoreDocs) {
        val shard = scoreDoc.shardIndex
        afters(shard) = Some(scoreDoc)
        val document = searchers(shard).doc(scoreDoc.doc, fields)
        documents.add((document, scoreDoc))
      }

      tracer.trace(s"Lucene index fetches $numFetched documents")
      logger.debug(s"Page fetched with $numFetched documents in $fetchTime")

    } catch {
      case e: Exception =>
        close()
        throw new IndexException(e, s"Error searching with $query and $sort")
    }
    if (finished) close()
  }

  /** Returns if more documents should be fetched from the Lucene index.
    *
    * @return `true` if more documents should be fetched, `false` otherwise
    */
  def needsFetch: Boolean = !finished && documents.isEmpty

  /** Returns if the iteration has more documents.
    *
    * @return `true` if the iteration has more documents, `false` otherwise
    */
  override def hasNext: Boolean = {
    if (needsFetch) fetch()
    !documents.isEmpty
  }

  /** Returns the next document-score tuple in the iteration.
    *
    * @return the next document-score tuple
    * @throws NoSuchElementException if the iteration has no more documents
    */
  override def next: (Document, ScoreDoc) = {
    if (hasNext) documents.poll else throw new NoSuchElementException
  }

  /** Closes the [[IndexSearcher]] and any other resources */
  override def close(): Unit = {
    if (!closed) try releaseSearchers() finally closed = true
  }

}

/** Companion object for [[DocumentIterator]]. */
object DocumentIterator {

  /** The max number of rows to be read per iteration. */
  val MAX_PAGE_SIZE = 10000
}
